shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -64,16 +65,13 @@
 @Slf4j
 public class CommandEncoder extends MessageToByteEncoder<Object> {
     private static final byte[] LENGTH_PLACEHOLDER = new byte[4];
-    private static final long RESET_TOKEN = 0;
-
     private final Function<Long, AppendBatchSizeTracker> appendTracker;
     private final Map<Map.Entry<String, UUID>, Session> setupSegments = new HashMap<>();
-    private long currentToken = RESET_TOKEN;
+    private AtomicLong tokenCounter = new AtomicLong(0);
```

Comment:
I don't understand why we need this token counter. To start with, why are we introducing this concept of token here?

Reply:
This is because of connection pooling;  
whenever we start append.. we generate the token give to timer which is set to execute after append batch time out. when time out happens then we check are we at the same counter; if not then no need to write any data because its already written as part of append breaks.
if the same token number exists , then issue the blocktimeout msg to enoder of the CommandEncoder.

Reply:
earlier.. (before connection pooling) .. there was dedicated channel segments writers , block time out used to check with current block size only (even that had issues) .. its all fixed now.